examples/java/src/main/java/org/apache/beam/examples/WindowedWordCount.java

```diff
@@ -211,19 +178,61 @@ public static void main(String[] args) throws IOException {
     PCollection<KV<String, Long>> wordCounts = windowedWords.apply(new WordCount.CountWords());
 
     /**
-     * Concept #5: Format the results for a BigQuery table, then write to BigQuery.
-     * The BigQuery output source supports both bounded and unbounded data.
+     * Concept #5: Key by the window, so we can write one file per window. To access the window in a
+     * {@link DoFn}, add a {@link BoundedWindow} parameter. This will be automatically detected and
+     * populated with the window for the current element.
      */
-    wordCounts.apply(ParDo.of(new FormatAsTableRowFn()))
-        .apply(BigQueryIO.Write
-          .to(getTableReference(options))
-          .withSchema(getSchema())
-          .withCreateDisposition(BigQueryIO.Write.CreateDisposition.CREATE_IF_NEEDED)
-          .withWriteDisposition(BigQueryIO.Write.WriteDisposition.WRITE_APPEND));
+    pipeline.getCoderRegistry().registerCoder(IntervalWindow.class, IntervalWindow.getCoder());
+    PCollection<KV<IntervalWindow, KV<String, Long>>> keyedByWindow =
+        wordCounts.apply(
+            ParDo.of(
+                new DoFn<KV<String, Long>, KV<IntervalWindow, KV<String, Long>>>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext context, BoundedWindow window) {
+                    context.output(KV.of((IntervalWindow) window, context.element()));
+                  }
+                }));
 
-    PipelineResult result = pipeline.run();
+    /**
+     * Concept #6: Format the results and write to a sharded file partitioned by window, using a
+     * simple ParDo operation. Because there may be failures followed by retries, the writes must be
+     * idempotent, and the files may not appear atomically.
+     */
+    keyedByWindow
+        .apply(GroupByKey.<IntervalWindow, KV<String, Long>>create())
+        .apply(
+            ParDo.of(
+                new DoFn<KV<IntervalWindow, Iterable<KV<String, Long>>>, Void>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext context) throws Exception {
+                    // Build a file name from the window
+                    DateTimeFormatter formatter = ISODateTimeFormat.dateTimeNoMillis();
```

Comment:
This is gross. Could certainly be pulled out and commented to be less gross. Probably should glance at logic to check I didn't botch something, too.